src/Development/IDE/Core/Shake.hs

```diff
@@ -565,86 +554,74 @@ shakeRestart IdeState{..} systemActs =
-          (,()) <$> newSession shakeExtras shakeDb dai cancelled)
+        (\() -> do
+          (,()) <$> newSession shakeExtras shakeDb acts)
 
 -- | Enqueue an action in the existing 'ShakeSession'.
 --   Returns a computation to block until the action is run, propagating exceptions.
 --   Assumes a 'ShakeSession' is available.
 --
 --   Appropriate for user actions other than edits.
-shakeEnqueue :: IdeState -> DelayedAction a -> IO (IO a)
-shakeEnqueue IdeState{shakeSession} act = shakeEnqueueSession shakeSession act
-
-shakeEnqueueSession :: MVar ShakeSession -> DelayedAction a -> IO (IO a)
-shakeEnqueueSession sess act = withMVar sess $ \s -> runInShakeSession s act
-
--- | Set up a new 'ShakeSession' with a set of initial system and user actions
--- Will crash if there is an existing 'ShakeSession' running.
--- Progress is reported only on the system actions.
--- Only user actions will get re-enqueued
-newSession :: ShakeExtras -> ShakeDatabase -> [DelayedActionInternal] -> [DelayedActionInternal] -> IO ShakeSession
-newSession ShakeExtras{..} shakeDb systemActs userActs = do
-    -- A work queue for actions added via 'runInShakeSession'
-    actionQueue :: TQueue DelayedActionInternal <- atomically $ do
-        q <- newTQueue
-        traverse_ (writeTQueue q) userActs
-        return q
-    actionInProgress :: TVar (Maybe DelayedActionInternal) <- newTVarIO Nothing
-
+shakeEnqueue :: ShakeExtras -> DelayedAction a -> IO (IO a)
+shakeEnqueue ShakeExtras{actionQueue} act = do
+    (b, dai) <- instantiateDelayedAction act
+    atomically $ pushQueue dai actionQueue
+    let wait' b =
+            waitBarrier b `catch` \BlockedIndefinitelyOnMVar ->
+                fail $ "internal bug: forever blocked on MVar for " <>
+                        actionName act
+    return (wait' b >>= either throwIO return)
+
+-- | Set up a new 'ShakeSession' with a set of initial actions
+--   Will crash if there is an existing 'ShakeSession' running.
+newSession :: ShakeExtras -> ShakeDatabase -> [DelayedActionInternal] -> IO ShakeSession
+newSession ShakeExtras{..} shakeDb acts = do
+    reenqueued <- atomically $ peekInProgress actionQueue
     let
         -- A daemon-like action used to inject additional work
         -- Runs actions from the work queue sequentially
-        pumpAction =
-            forever $ do
-                join $ liftIO $ atomically $ do
-                    act <- readTQueue actionQueue
-                    writeTVar actionInProgress $ Just act
-                    return (logDelayedAction logger act)
-                liftIO $ atomically $ writeTVar actionInProgress Nothing
+        pumpActionThread = do
+            d <- liftIO $ atomically $ popQueue actionQueue
+            void $ parallel [run d, pumpActionThread]
+
+        run d  = do
+            start <- liftIO offsetTime
+            getAction d
+            liftIO $ atomically $ doneQueue d actionQueue
```

Comment:
There can be an exception after `getAction d` but before `doneQueue`. Should the `popQueue`, `getAction` and `doneQueue` calls be wrapped in `bracket`?

Reply:
Ok, so `bracket` won't be appropriate since we need to requeue, but perhaps we can `mask` exceptions everywhere but in `getAction d`.

Reply:
I'm tempted to leave as it is. The outcome of not calling doneQueue is just redundant work: the action will be reexecuted in the next restart, then the check in line 630 will not signal the barrier, and finally the action will be removed from the queue

Reply:
OK, makes sense.